File: Excel/514 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/514 Sub Grid Maximum Sum.xlsx"
input = read_excel(path, range = "B2:F6", col_names = F) %>% as.matrix()
test  = read_excel(path, range = "H1:H4") %>% arrange(`Answer Expected`)

indices = expand.grid(i = 1:(nrow(input) - 1), j = 1:(ncol(input) - 1))
results = indices %>%
  pmap(function(i, j) {
    sub_mat = input[i:(i + 1), j:(j + 1)]
    list(matrix = sub_mat, sum = sum(sub_mat, na.rm = TRUE))
  })
max_sum = max(map_dbl(results, "sum"))
max_subs = keep(results, ~ .x$sum == max_sum)
max_subs_str = map_chr(max_subs, ~ paste(apply(.x$matrix, 1, paste, collapse = ", "), collapse = " ; ")) %>%
  tibble(`Answer Expected` = .) %>%
  arrange(`Answer Expected`)

identical(max_subs_str, test)
#> [1] TRUE

File: Excel/500-599/514/514 Challenge.py
import pandas as pd
import numpy as np
import itertools

path = "514 Sub Grid Maximum Sum.xlsx"
input = pd.read_excel(path, usecols="B:F", header=None, skiprows=1)
input_mat = input.to_numpy()
test = pd.read_excel(path, usecols="H", nrows = 3)

indices = list(itertools.product(range(input_mat.shape[0] - 1), range(input_mat.shape[1] - 1)))

results = []
for i, j in indices:
    sub_matrix = input_mat[i:i + 2, j:j + 2]
    sub_sum = np.nansum(sub_matrix)
    results.append({'matrix': sub_matrix, 'sum': sub_sum})

max_sum = max(res['sum'] for res in results)
max_subs = [res for res in results if res['sum'] == max_sum]

max_subs_str = [' ; '.join(', '.join(map(str, row)) for row in res['matrix']) for res in max_subs]
max_subs_df = pd.DataFrame({'Answer Expected': max_subs_str})


print(max_subs_df.equals(test)) # True